### PR TITLE
Adding Scans for Deprecated Function in Core

### PR DESCRIPTION
**Pull request by aweingarten**
This PR will scan custom code in the standard locations for deprecated functions and fail the build if they are found.  This will help the progression to D9! 

It should the same patterns as PHPCS or Symfony Security Scans

### CODE REVIEW

**Comment by mikemadison13** [comment]
@aweingarten might make sense to add this to the validate:all (similar to how we did the security one) so it runs as part of the standard validation process, then we can disable parts of it.

**Comment by mikemadison13** [comment]
also, IMHO, this should be disabled by default.

File: src/Robo/Commands/Validate/PHPStanCommand.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Converts fileset to a string of all the files.
+     *
+     * @param \Symfony\Component\Finder\Finder[] $filesets
+     *
+     * @return string | NULL
+     *   A space delimeted string of files.
+     */
+    protected function fileSetToList(array $filesets) {
```
Comment (aweingarten) [inline src/Robo/Commands/Validate/PHPStanCommand.php]: Does this make sense as a more global helper function?

**Comment by aweingarten** [comment]
1: Added this to the validate all call.
2: Dumb question how do I disable this off by default?  Is that in the build.yml file? 
like by adding `disable-targets.tests:phpstan`

**Comment by mikemadison13** [comment]
@aweingarten i'm ok with it running, as long as it's not scanning files. so maybe just comment out the file sets by default?

**Comment by mikemadison13** [comment]
FYI looks like phpcs validation failures on the build failure

**Comment by aweingarten** [comment]
PHPCS addressed

**Comment by aweingarten** [comment]
@mikemadison13 okay I think all the feedback has been addressed.  How do you want this to be socialized/documented?

**Comment by mikemadison13** [comment]
honestly, we don't have a section on validation in the docs (that i see). we probably need to refactor the testing docs to better differentiate between automated testing and validation that occurs in blt. let me think on this a bit and get back to you.

**Comment by aweingarten** [comment]
Sounds good.  I am thinking that we should also probably think about adding it to the the release notes so that people start enabling it.  

Also once we get this turned on for deprecation, increasing the sensitivity of PHPstan to Level 2-7 scans is also freaking awesome.  It's really worth considering nudging people as we head towards D9. Checkout the results of the scan for Acquia connector.  

[acquia-connector-scan.txt](https://github.com/acquia/blt/files/2871187/acquia-connector-scan.txt)

**Comment by aweingarten** [comment]
So @mikemadison13 I'm an idiot.  There is an obvious answer here.  What if we make this work like PHPCS where we only scan the files that have _changed_ rather than all the things?  That way we gently make the world a better place without immediate whiplash?

**Comment by mikemadison13** [comment]
@aweingarten i really like that idea, although i would be interested in benchmarking it. i do think that creates some disparity between "why are my changed files getting deprecation scans" and my others ones aren't.

i do think we need this to be on globally or off globally, so maybe the commit scan needs to respect the same rules as the normal scan (just within the changed fileset)

**Comment by aweingarten** [comment]
So anecdotally the time for our custom files is under 30 seconds.  As for the disparity it already exists for PHPCS so I think we are okay.

**Comment by mikemadison13** [comment]
it does, but re: https://github.com/acquia/blt/issues/3366 i'm trying to change that since commit hooks find different things than blt validate and that is driving me batty.

File: composer.json [hunk 2]
```diff
@@ -50,7 +50,10 @@
         "symfony/yaml": "^3.2.8",
         "tivie/php-os-detector": "^1.0",
         "typhonius/acquia-php-sdk-v2": "^1.0.3",
-        "zaporylie/composer-drupal-optimizations": "^1.0"
+        "zaporylie/composer-drupal-optimizations": "^1.0",
```
Comment (aweingarten) [inline composer.json]: Reorder this one alphabetically with the rest of the packages

File: template/phpstan.neon [hunk 3]
```diff
@@ -0,0 +1,18 @@
+parameters:
+  customRulesetUsed: true
+  memory_limit: -1
```
Comment (aweingarten) [inline template/phpstan.neon]: Probably set this to a more sane value

File: template/phpstan.neon [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  reportUnmatchedIgnoredErrors: false
+  # Ignore phpstan-drupal extension's rules.
+  excludes_analyse:
+  - blazy_test.module
```
Comment (aweingarten) [inline template/phpstan.neon]: Add comment here for why

File: config/build.yml [hunk 5]
```diff
@@ -225,6 +225,11 @@ validate:
     filters: { }
     # Add any custom Twig functions for linter to ignore.
     functions: { }
+  phpstan:
```
Comment (aweingarten) [inline config/build.yml]: Clean up the commit message here.

File: src/Robo/Commands/Validate/PHPStanCommand.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace Acquia\Blt\Robo\Commands\Validate;
+
+use Acquia\Blt\Robo\BltTasks;
+use Acquia\Blt\Robo\Exceptions\BltException;
+
+/**
+ * Defines commands in the "validate:*" namespace.
+ */
+class PHPStanCommand extends BltTasks {
```
Comment (aweingarten) [inline src/Robo/Commands/Validate/PHPStanCommand.php]: Capitalization should match the file

**Comment by alexxed** [comment]
What if on the first run, you generate exclusions based on current state of things so that the developer can work on that in his own time, but have phpstan run for any new code he adds. You can use ignoreErrors in phpstan.neon with regex and the filename to be exact.

On the next run, just show a single warning that there are ignoreErrors to be dealt with.

**Comment by aweingarten** [comment]
@alexxed that's a really interesting idea. 
That being said we solved this years ago when we started using phpcs.  Back then most of our codebases bombed out on PHPCS.  It was gloriously bad. The solution back then, which is what we do today, was by default to only scan the files that were changed on commit or in a PR.  

It's a `you touch it you own it and need to make it better` approach.   BLT users have already been trained on this and it will come naturally to them.

File: tests/phpunit/Blt/PhpStanTest.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ */
+class PhpStanTest extends BltProjectTestBase {
+
+  /**
+   * @group blt
+   *
+   * @dataProvider providerPhpStanFilesBootstrapProvider
+   */
+  public function testPhpStanFilesBootstrap($filename, $needle, $contains) {
+    $process = new Process("./vendor/bin/phpstan analyse $filename --configuration=phpstan.neon");
```
Comment (grasmash) [inline tests/phpunit/Blt/PhpStanTest.php]: We need to run the actual BLT command in testing.

File: composer.json [hunk 8]
```diff
@@ -55,10 +55,13 @@
...
         "knplabs/github-api": "^2.6",
+        "mglaman/phpstan-drupal": "^0.11.1",
         "php-http/guzzle6-adapter": "^1.1",
-        "webflo/drupal-core-require-dev": "^8.6.0",
+        "phpstan/phpstan": "^0.11.1",
+        "phpstan/phpstan-deprecation-rules": "^0.11.0",
         "phpunit/phpunit": "^4.8|^6.5",
-        "squizlabs/php_codesniffer": "^3.0.1"
+        "squizlabs/php_codesniffer": "^3.0.1",
+        "webflo/drupal-core-require-dev": "^8.6.0"
```
Comment (grasmash) [inline composer.json]: Why is this necessary?

File: src/Robo/Commands/Validate/PHPStanCommand.php [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+   * if those files are a subset of the phpstan.filesets filesets.
+   *
+   * @command tests:phpstan:sniff:modified
+   * @aliases tpsm
+   *
+   * @return int
+   */
+  public function sniffModified() {
+    $this->say("Sniffing modified and untracked files in repo...");
+    $arguments = "--filter=gitmodified " . $this->getConfigValue('repo.root');
```
Comment (grasmash) [inline src/Robo/Commands/Validate/PHPStanCommand.php]: `--filter=gitmodified`

I'm a bit concerned this may not catch exactly the files we care about. Can we test that the right files  are scanned?

**Comment by danepowell** [comment]
The disconnect between pre-commit validation and `blt validate` is jarring enough already, I don't want to add this in a way that tests might pass `blt validate` but fail during pre-commit.

Rather, I'd like to see this run alongside all of the other validations during `blt validate`.

In 10.0.x, it can be enabled by default, _if_ we can get this PR in before 10.0.0 drops. But it should have a message clearly indicating how to disable it when/if it fails.

In 9.2.x, it should be disabled by default (if we decide to backport it to 9.2.x at all).

**Comment by aweingarten** [comment]
@danepowell
Thanks for the feedback!  I added some info to the error message for how to turn off the scans! 

> The disconnect between pre-commit validation and blt validate is jarring enough already, I don't want to add this in a way that tests might pass blt validate but fail during pre-commit.  Rather, I'd like to see this run alongside all of the other validations during blt validate.

Can you elaborate on this?  Let me know what you are looking for here.  _I thought_  that the phpstan scans were built into `blt validate`.   So the PR is intended to replicate the PHPCS behavior.  Are you requesting that we also refactor PHPCS functionality in this PR?

File: composer.json [hunk 10]
```diff
@@ -55,10 +55,12 @@
...
         "composer/composer": "^1.6.4",
         "knplabs/github-api": "^2.6",
+        "mglaman/phpstan-drupal": "^0.11.1",
         "php-http/guzzle6-adapter": "^1.1",
-        "webflo/drupal-core-require-dev": "^8.6.0",
+        "phpstan/phpstan": "^0.11.1",
+        "phpstan/phpstan-deprecation-rules": "^0.11.0",
         "phpunit/phpunit": "^4.8|^6.5",
-        "squizlabs/php_codesniffer": "^3.0.1"
+        "squizlabs/php_codesniffer": "^3.0.1",
```
Comment (danepowell) [inline composer.json]: Think this trailing comma isn't valid JSON

**Comment by danepowell** [comment]
The problem with PHPCS is that if you pass it a list of files to check (i.e. during git pre-commit hook), it will _always_ scan those files even if they wouldn't otherwise be scanned according to the defined filesets in BLT and/or phpcs.xml.dist.

As long as this filters the file list (i.e. scans an intersection between git changed files and the phpstan fileset) it should be fine.

**Comment by danepowell** [comment]
Looks like tests are still failing, and there are still some comments from Matt to address.

**Comment by aweingarten** [comment]
@danepowell I should have time this weekend to devote to this PR!

**Comment by aweingarten** [comment]
Closing given the updated direction in BLT unsure if this is an appropriate place to commit this.
